src/applications/bmqstoragetool/m_bmqstoragetool_cslprinter.cpp

```diff
@@ -0,0 +1,568 @@
+            d_ostream << '\n' << queueInfosSize << " Queues found:\n";
+            bsl::vector<bmqp_ctrlmsg::QueueInfo>::const_iterator itEnd =
+                queueInfos.cend();
+            if (queueInfosSize > queuesLimit) {
+                d_ostream << "Only first " << queuesLimit
+                          << " queues are displayed.\n";
+                itEnd = queueInfos.cbegin() + queuesLimit;
+            }
+            bsl::vector<bmqp_ctrlmsg::QueueInfo>::const_iterator it =
+                queueInfos.cbegin();
+            for (; it != itEnd; ++it) {
+                d_ostream << *it << '\n';
+            }
+        }
+    }
+
+    void printFooter(const CslRecordCount& recordCount,
+                     const Parameters::ProcessCslRecordTypes&
+                         processCslRecordTypes) const BSLS_KEYWORD_OVERRIDE
+    {
+        if (processCslRecordTypes.d_snapshot) {
+            recordCount.d_snapshotCount > 0
+                ? (d_ostream << recordCount.d_snapshotCount << " snapshot")
+                : d_ostream << "No snapshot";
+            d_ostream << " record(s) found.\n";
+        }
+        if (processCslRecordTypes.d_update) {
+            recordCount.d_updateCount > 0
+                ? (d_ostream << recordCount.d_updateCount << " update")
+                : d_ostream << "No update";
+            d_ostream << " record(s) found.\n";
+        }
+        if (processCslRecordTypes.d_commit) {
+            recordCount.d_commitCount > 0
+                ? (d_ostream << recordCount.d_commitCount << " commit")
+                : d_ostream << "No commit";
+            d_ostream << " record(s) found.\n";
+        }
+        if (processCslRecordTypes.d_ack) {
+            recordCount.d_ackCount > 0
+                ? (d_ostream << recordCount.d_ackCount << " ack")
+                : d_ostream << "No ack";
+            d_ostream << " record(s) found.\n";
+        }
+    }
+};
+
+// ====================
+// class JsonCslPrinter
+// ====================
+
+class JsonCslPrinter : public CslPrinter {
+  protected:
+    // PROTECTED DATA
+
+    bsl::ostream&     d_ostream;
+    bslma::Allocator* d_allocator_p;
+    mutable bool      d_braceOpen;
+    mutable bool      d_firstRow;
+
+    // PROTECTED METHODS
+
+    void openBraceIfNotOpen(const std::string& fieldName) const
+    {
+        if (!d_braceOpen) {
+            d_ostream << "  \"" << fieldName << "\": [\n";
+            d_braceOpen = true;
+        }
+        else {
+            d_ostream << ",\n";
+        }
+    }
+
+    void closeBraceIfOpen() const
+    {
+        if (d_braceOpen) {
+            d_ostream << "\n  ]";
+            d_braceOpen = false;
+            d_firstRow  = false;
+        }
+        if (!d_firstRow) {
+            d_ostream << ",\n";
+        }
+        else {
+            d_firstRow = false;
+        }
+    }
+
+  public:
+    // CREATORS
+
+    JsonCslPrinter(bsl::ostream& os, bslma::Allocator* allocator)
```

Comment:
```suggestion
    explicit JsonCslPrinter(bsl::ostream& os, bslma::Allocator* allocator = 0)
```
+ nested trait uses bslma allocator